sdk/core/azure-core/src/main/java/com/azure/core/util/implementation/ReferenceManagerImpl.java

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util.implementation;
+
+import com.azure.core.util.ReferenceManager;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Implementation of {@link ReferenceManager}.
+ */
+public final class ReferenceManagerImpl implements ReferenceManager, Runnable {
```

Comment:
I'm not always a super-big fan of having classes implement `Runnable` like this, because it tends to conflate some 'outer' functionality with some 'inner' functionality. It took me a bit of time to identify where the `run()` was being called - I instinctively looked into the `CleanableReference` code first, and started to wonder if there was a bug, before I realised the new Thread was being given `this` to call it. Often what I would prefer to do is this (after removing `Runnable` from the class and renaming `run` to something like `clearReferenceQueue`):

```java
Thread thread = new Thread(Thread.currentThread().getThreadGroup(), this::clearReferenceQueue,
            BASE_THREAD_NAME + RESOURCE_MANAGER_THREAD_NUMBER.getAndIncrement());
```